kion/resource_aws_account.go

```diff
@@ -208,6 +208,90 @@ func resourceAwsAccount() *schema.Resource {
+	}
+
+	// Send the POST request to create the AWS account.
+	respCache, err := c.POST("/v3/account-cache/create?account-type=aws", postCacheData)
+	if err != nil || respCache.RecordID == 0 {
+		if err == nil {
+			err = fmt.Errorf("received item ID of 0")
+		}
+		return diag.Errorf("Unable to create AWS Account: %v", err), 0
+	}
+
+	// Wait for the account to be fully created.
+	if err := waitForAccountCreation(c, ctx, respCache.RecordID, d); err != nil {
+		return diag.FromErr(err), 0
+	}
+
+	// Return any diagnostics and the ID of the created account cache.
+	return diags, respCache.RecordID
+}
+
+// populateOrgUnitFromResourceData parses OU details from Terraform data, updating AccountCacheNewAWSCreate for account creation.
+func populateOrgUnitFromResourceData(c *hc.Client, postCacheData *hc.AccountCacheNewAWSCreate, d *schema.ResourceData) error {
+	if v, exists := d.GetOk("aws_organizational_unit"); exists {
+		orgUnitSet := v.(*schema.Set)
+		for _, item := range orgUnitSet.List() {
+			orgUnitMap, ok := item.(map[string]interface{})
+			if !ok {
+				return fmt.Errorf("invalid format for aws_organizational_unit")
+			}
+			postCacheData.OrganizationalUnit = &hc.PayerOrganizationalUnit{
+				Name:      orgUnitMap["name"].(string),
+				OrgUnitId: orgUnitMap["org_unit_id"].(string),
+			}
+		}
+	}
+	return nil
+}
+
+// waitForAccountCreation polls the creation status until the account is created or a timeout occurs.
+func waitForAccountCreation(c *hc.Client, ctx context.Context, accountCacheId int, d *schema.ResourceData) error {
+	createStateConf := &resource.StateChangeConf{
+		// Define the refresh function, which checks the account creation status.
+		Refresh: func() (interface{}, string, error) {
+			resp := new(hc.AccountResponse)
+			err := c.GET(fmt.Sprintf("/v3/account-cache/%d", accountCacheId), resp)
+			if err != nil {
+				// Handle the case where the account is not found as a transient state, not an error.
+				if resErr, ok := err.(*hc.RequestError); ok && resErr.StatusCode == http.StatusNotFound {
+					tflog.Trace(ctx, fmt.Sprintf("Checking new AWS account status: /v3/account-cache/%d not found", accountCacheId), map[string]interface{}{"accountCacheId": accountCacheId})
+					return nil, "NotFound", nil
+				}
+				tflog.Trace(ctx, fmt.Sprintf("Checking new AWS account status: /v3/account-cache/%d error", accountCacheId), map[string]interface{}{"error": err, "accountCacheId": accountCacheId})
+				return nil, "Error", err
+			}
+
+			// Check if the account number is still not available in the response.
+			if resp.Data.AccountNumber == "" {
+				tflog.Trace(ctx, fmt.Sprintf("Checking new AWS account status: /v3/account-cache/%d missing account number", accountCacheId), map[string]interface{}{"accountCacheId": accountCacheId})
+				return resp, "MissingAccountNumber", nil
+			}
+
+			// Account creation is successful.
+			return resp, "AccountCreated", nil
+		},
+		Pending: []string{"MissingAccountNumber", "NotFound"},
```

Comment:
You should not need to add `NotFound` here.  The plugin framework handles not found stuff for you: https://github.com/hashicorp/terraform-plugin-sdk/blob/v2.30.0/helper/retry/state.go#L136-L147. And it's usually better if you let it handle not founds instead of treating this as pending because it has different behavior between the two.

Reply:
Thanks for this.  TIL.